ftlib/commlib/nccl/impl.py

```diff
@@ -36,75 +35,92 @@ class NCCL(BasicCommLib):
+        if timeout is None:
+            timeout == self._default_timeout
 
-        success = self._nccl_context.setInput(data)
-        if not success:
-            raise Exception("set input failed")
+        logging.debug("broadcasting: " + str(data))
 
-        success = self._nccl_context.broadcastAsync(root)
-        if not success:
-            raise Exception("async broadcast operation failed")
+        if not isinstance(data, np.ndarray):
+            data = np.array(data).astype(np.float32)
 
-        signal.alarm(10)
+        if data.dtype is not np.float32:
+            raise NotImplementedError(
+                "data types other fp32 are not implemented"
+            )
 
-        success = False
-        while not success:
-            success = self._nccl_context.checkOpResult(1)
+        broadcast_call = self._nccl_context.broadcast(data, root_rank)
+        signal.alarm(timeout)
+        call_success = False
+        while not call_success:
+            call_success = broadcast_call.check_complete()
         signal.alarm(0)
 
-        data = self._nccl_context.getOutput()
+        if not call_success:
+            raise RuntimeError("broadcast failed")
+
         logging.debug("receiving: " + str(data))
 
+        # the data is broadcast in-place
+        return data
+
     @BasicCommLib.register_api
-    def allreduce(self, data, op="SUM"):
+    def allreduce(self, data, op="SUM", timeout=None):
         if op != "SUM":
             raise ValueError(
                 "Only SUM operation is currently allowed in NCCL allreduce"
             )
+        if timeout is None:
+            timeout = self._default_timeout
         logging.debug("averaging: " + str(data))
 
-        success = self._nccl_context.setInput(data)
-        if not success:
-            raise Exception("set input failed")
+        if not isinstance(data, np.ndarray):
+            data = np.array(data).astype(np.float32)
 
-        success = self._nccl_context.allreduceAsync()
-        if not success:
-            raise Exception("asynchronizely allreduce failed")
-
-        signal.alarm(10)
+        if data.dtype is not np.float32:
+            raise NotImplementedError(
+                "data types other fp32 are not implemented"
+            )
 
-        success = False
-        while not success:
-            success = self._nccl_context.checkOpResult(1)
+        allreduce_call = self._nccl_context.allreduce(data)
+        signal.alarm(timeout)
+        call_success = False
+        while not call_success:
+            call_success = allreduce_call.check_complete()
         signal.alarm(0)
 
-        output = self._nccl_context.getOutput()
-        logging.debug("receiving: " + str(output))
+        if not call_success:
+            raise RuntimeError("allreduce failed")
+
+        logging.debug("receiving: " + str(data))
+
+        # the data is allreduced in-place
+        return data
+
+    def barrier(self):
+        # maybe we can allreduce a byte
```

Comment:
We need to be consistent with other commlib implementations though.

Reply:
Sure. I'll add script here.